examples/shopping_cart/src/main.rs

```diff
@@ -0,0 +1,129 @@
+// This is an example of using pre/post conditions as well as invariants
+// via the contracts crate.
+//
+// We define a simple shopping cart with an invariant both on items in the cart
+// and the overall cart.
+//
+// TODO(wrwg): this example is not yet fully functional regarding MIRAI verification.
+use contracts::*;
+
+use mirai_annotations::*;
+
+// Items stored in the shopping cart.
+pub struct Item {
+    // TODO(wrwg): make this String once supported
+    name: &'static str,
+    price: f64,
+}
+
+impl Item {
+    // Invariant for the Item struct. We define those invariants via functions
+    // so they can easily be referenced from attributes.
+    fn invariant(&self) -> bool {
+        self.name.len() > 0 && self.price > 0.0
+    }
+
+    // Creates a new Item, satisfying the invariant. Notice we can't use
+    // the `#[invariant]` attribute for new because this requires a `self` argument.
+    #[post(ret.invariant())]
```

Comment:
The post condition implies a precondition, but I would prefer to see the precondition spelled out in a pre.

Reply:
Did it, but this is actually an interesting case which I expect to be quite common, and which needs a better solution. We essentially attaching a pre-condition to a pure data constructor which is identical with the invariant. If invariants become more complex, we don't want the author to need to duplicate the predicates. 

What I really would like to do here is:

    #[invariant(P)] 
    impl S {
      // Because this returns S, invariant should be applied to the result.
      // However, currently invariants are only applied to self arguments.
       fn new(params...) -> S { ... }
       ...
    }

I guess I open an issue for this on the contracts crate.